ampproject/amp-wp
# JavaScript Tests, Take 1.

This PR aims to make the current JavaScript code base more testable, while also adding some initial tests to prove that the setup is working. There's still lots to do. Mainly:

* [x] Test and verify webpack configuration  
  This PR also affects non-Stories components, so we need to double-check whether it works and makes sense.
* [x] Maybe reorganize components some more.  
  Ideally, shared components like the featured image notice stuff is in a common directory, instead of in only `amp-stories` or `block-editor`. See also #2038.
* [x] Make sure Jest works (!)
  There are currently some issues, see https://github.com/WordPress/gutenberg/pull/15139, https://wordpress.slack.com/archives/C02QB2JS7/p1556031641343900, https://github.com/WordPress/gutenberg-examples/issues/26.  
  Might need a second pair of eyes from Gutenberg devs.

**Testing Instructions**

The two main things to test here are:

1. Make sure everything in the admin still works: toggling error details, block validation, the stories editor. (Manual testing)
2. Ensuring the unit tests are running. (Run `npm run test:js`)

Note: right now, the unit test part is a bit more complex at the moment because the new Gutenberg packages haven't been published on npm yet.

In order to prevent any errors and make sure that all the tests run smoothly, you need to perform the following tasks:

1. Clone https://github.com/WordPress/gutenberg/ and build it.
2. Go to each of these folders and run `npm link` inside them:  
    * `packages/babel-preset-default`
    * `packages/block-editor`
    * `packages/blocks`
    * `packages/compose`
    * `packages/data`
    * `packages/element`
    * `packages/hooks`
    * `packages/i18n`
    * `packages/wordcount`
   
   So basically this: `(cd packages/compose && npm link); (cd packages/blocks && npm link); ...`)
3. Check out this PR branch locally
4. Run `npm link` for all these packages.  
   Basically this: `npm link @wordpress/compose; npm link @wordpress/blocks; npm link @wordpress/i18n; ...`
5. Finally, run `npm run test:js`

The result should look like this:

```
Test Suites: 1 failed, 18 passed, 19 total
Tests:       5 failed, 56 passed, 61 total
Snapshots:   1 file removed, 6 passed, 6 total
Time:        25.028s, estimated 36s
```

(Yes, there are currently some failing tests; see below for the reason)

**To do** (ideally in some follow-up PRs)

* [ ] See how to best mock `select()` calls in tests as the current approach does not seem to be working in the `test/reducer.js`.
* [ ] Add some more unit tests where applicable  
  This is more of an ongoing process
* [ ] Refine E2E test setup and add some initial E2E tests
* [ ] Run tests on Travis  
  Once all tests are green, otherwise it's just gonna be a sad experience 🙂 

See #2121.
